### ISSUE TITLE
Bug: `unsupported-features/node-builtins-modules` doesn't work for semver range in v17

### ISSUE THREAD

**Issue by teppeis** [eslint-community/eslint-plugin-n#250]
### Environment

Node version: v20.12.2
npm version: 10.5.0
ESLint version: 8.57.0
eslint-plugin-n version: 17.2.1
Operating System: macOS

### What rule do you want to report?

unsupported-features/node-builtins-modules

### Link to Minimal Reproducible Example

https://eslint-online-playground.netlify.app/#eNpdkl1PwyAUhv8K4Woma7cmxmnNduW9N8abMROkx6WzBTzAnNn63z2U7pMr+nJ4eDh0zx2qCexkaxvIN46XXBntPJPOAXo2Zwg/oUYYCZ4iwe+ehdbwO5TkCBtQ3o1izMc8B9fU2qMimtHE2wvNmOCw86ArJ3jJlvTVF5UIyrQt5VAJPqYq24R1rUs9uVpZjRPDSqQjX62vyTGSenZPV618B3S0EHPBG+nBkWy/M1Y4E1DB25+FVNCaKjQ0j+vdwEdKLrlpCK4n2mRBu2CtQQ9V9gXSBwRHeQXZZ6gbXyej5XknWSEaPDnEcQPeXiov5k/5NJ8mpTS643TVawrdUYetVN9yHV/rqr+aGndlT95kl9gfxX1ezPIpOxzYYl48nA863r2C7QvY2HCt6htOeq1Eerzoacqz9GjZcI1iNoDJlv6I7h9Tur5F

### What did you expect to happen?

This works fine (expectedly fail),
```js
"n/no-unsupported-features/node-builtins": ["error", {version: "9.0.0"}]
```
but this doesn't work (unexpectedly success) in v17: [demo](https://eslint-online-playground.netlify.app/#eNpdkl1PwyAUhv8K4Woma7cmxmnNduW9N8abMROkx6WzBTzAnNn63z2U7pMr+nJ4eDh0zx2qCexkaxvIN46XXBntPJPOAXo2Zwg/oUYYCZ4iwe+ehdbwO5TkCBtQ3o1izMc8B9fU2qMimtHE2wvNmOCw86ArJ3jJlvTVF5UIyrQt5VAJPqYq24R1rUs9uVpZjRPDSqQjX62vyTGSenZPV618B3S0EHPBG+nBkWy/M1Y4E1DB25+FVNCaKjQ0j+vdwEdKLrlpCK4n2mRBu2CtQQ9V9gXSBwRHeQXZZ6gbXyej5XknWSEaPDnEcQPeXiov5k/5NJ8mpTS643TVawrdUYetVN9yHV/rqr+aGndlT95kl9gfxX1ezPIpOxzYYl48nA863r2C7QvY2HCt6htOeq1Eerzoacqz9GjZcI1iNoDJlv6I7h9Tur5F)
```js
"n/no-unsupported-features/node-builtins": ["error", {version: ">=9.0.0"}]
```

v16 works fine: [demo](https://eslint-online-playground.netlify.app/#eNpdkl1PwyAUhv8K4Woma7cmZmrNduW9N8abMROkx6WzBTzAnNn63z2U7pMr+nJ4eDh0zx2qCexkaxvIN46XXBntPJPOAXo2Zwg/oUYYCZ4iwe+ehdbwO5TkCBtQ3o1izMc8B9fU2qMimtHE2wvNmOCw86ArJ3jJlvTVF5UIyrQt5VAJPqYq24R1rUs9uVpZjRPDSqQjX62vyTGSenZPV618B3S0EHPBG+nBkWy/M1Y4E1DB25+FVNCaKjQ0j+vdwEdKLrlpCK4n2mRBu2CtQQ9V9gXSBwRHeQXZZ6gbXyej5XknWSEaPDnEcQPeXiov5k/5NJ8mpTS643TVawrdUYetVN9yHV/rqr+aGndlT95kl9gfxX1ePORTdjiwxbyYnQ863r2C7QvY2HCt6htOeq1Eerzoacqz9GjZcI1iNoDJlv6I7h9Tq75E)

### Participation

- [ ] I am willing to submit a pull request for this issue.

### Additional comments

Since specifying the version by range is common in the engines field of package.json, this bug is likely to occur in many cases.
However, this is not included in the test cases.

The logic of this range comparison seems to be the cause.
https://github.com/eslint-community/eslint-plugin-n/blob/eb11b5b35a6a797dc7fba6df53b1c4dada3a2a55/lib/util/check-unsupported-builtins.js#L40-L53

**Comment by scagood** [comment]
Thank you for this report!

We have two options here:
1. Convert the range to a minimum version
2. Throw an error to state that the config must not be a version

I think in this context that converting it to a minimum version is the best course of action.

What do you think @aladdin-add?

**Comment by teppeis** [comment]
> 2. Throw an error to state that the config must not be a version range

What about the `engines` field of package.json? Mostly range is used there.

> 1. Convert the range to a minimum version

When the engines field is `>=18.10.0 || >= 20.0.0`, it should be converted to `18.10.0` ? When supported versions are `["20.1.0", "18.9.0"]`, it cannot be determined whether it is supported using the minimum version.

It would be better to use the same logic as before.
https://github.com/eslint-community/eslint-plugin-n/blob/16.6.2/lib/util/check-unsupported-builtins.js#L54-L60

**Comment by scagood** [comment]
The same section you highlighted can be found here on master:
https://github.com/eslint-community/eslint-plugin-n/blob/master/lib/util/check-unsupported-builtins.js#L40-L47

**Comment by teppeis** [comment]
The difference:
- v16: `return !configured.intersects(unsupportedRange)` (`configured` is expected to be range)
- v17: `return configured.intersects(supportedRange)`　(`configured` is expected to be only minimum version)

**Comment by aladdin-add** [comment]
yes, the configured version should be a subset of the supported range. I have made a fix: #252

**Comment by scagood** [comment]
A fix was implemented in #252, but due to an [issue in semver](https://github.com/npm/node-semver/issues/703) we needed to revert it.

For now I wonder if we can solve this another way 🤔
